spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/repository/query/SpannerStatementQueryExecutor.java

```diff
@@ -169,32 +171,63 @@ private SpannerStatementQueryExecutor() {
-						+ childPersistentEntity.tableName() + " WHERE ");
-		StringJoiner sj = new StringJoiner(" and ");
+			SpannerPersistentEntity<T> childPersistentEntity, SpannerCustomConverter writeConverter,
+			SpannerMappingContext mappingContext) {
+		return getChildrenRowsQuery(KeySet.singleKey(parentKey), childPersistentEntity, writeConverter, mappingContext);
+	}
+
+	public static <T> Statement getChildrenRowsQuery(KeySet keySet,
+			SpannerPersistentEntity<T> childPersistentEntity, SpannerCustomConverter writeConverter,
+			SpannerMappingContext mappingContext) {
+		StringJoiner orJoiner = new StringJoiner(" OR ");
 		List<String> tags = new ArrayList<>();
 		List keyParts = new ArrayList();
 		int tagNum = 0;
 		List<SpannerPersistentProperty> childKeyProperties = childPersistentEntity
 				.getFlattenedPrimaryKeyProperties();
-		Iterator parentKeyParts = parentKey.getParts().iterator();
-		while (parentKeyParts.hasNext()) {
-			SpannerPersistentProperty keyProp = childKeyProperties.get(tagNum);
-			String tagName = "tag" + tagNum;
-			sj.add(keyProp.getColumnName() + " = @" + tagName);
-			tags.add(tagName);
-			keyParts.add(parentKeyParts.next());
-			tagNum++;
+
+		for (Key key : keySet.getKeys()) {
+			StringJoiner andJoiner = new StringJoiner(" AND ", "(", ")");
+			Iterator parentKeyParts = key.getParts().iterator();
+			while (parentKeyParts.hasNext()) {
+				SpannerPersistentProperty keyProp = childKeyProperties.get(tagNum % childKeyProperties.size());
+				String tagName = "tag" + tagNum;
+				andJoiner.add(keyProp.getColumnName() + " = @" + tagName);
+				tags.add(tagName);
+				keyParts.add(parentKeyParts.next());
+				tagNum++;
+			}
+			orJoiner.add(andJoiner.toString());
 		}
-		return buildStatementFromSqlWithArgs(sb.toString() + sj.toString(), tags, null, writeConverter,
+		String cond = orJoiner.toString();
+		String sb = "SELECT " + getColumnsStringForSelect(childPersistentEntity, mappingContext) + " FROM "
+				+ childPersistentEntity.tableName() + (cond.isEmpty() ? "" : " WHERE " + cond);
+		return buildStatementFromSqlWithArgs(sb, tags, null, writeConverter,
 				keyParts.toArray(), null);
 	}
 
+	private static <C, P> String getChildrenStructsQuery(
+			SpannerPersistentEntity<C> childPersistentEntity,
+			SpannerPersistentEntity<P> parentPersistentEntity, SpannerMappingContext mappingContext,
+			String columnName) {
+		String tableName = childPersistentEntity.tableName();
+		List<SpannerPersistentProperty> parentKeyProperties = parentPersistentEntity
+				.getFlattenedPrimaryKeyProperties();
+		String condition = parentKeyProperties.stream()
+				.map(keyProp -> tableName + "." + keyProp.getColumnName()
+						+ " = "
+						+ parentPersistentEntity.tableName() + "." + keyProp.getColumnName())
+				.collect(Collectors.joining(" AND "));
+
+		return "ARRAY (SELECT AS STRUCT " + getColumnsStringForSelect(childPersistentEntity, mappingContext) + " FROM "
```

Comment:
I wonder if it makes sense to pass string builders around as the query gets constructed instead of returning strings that have to, in turn, be concatenated. For a high-volume system, all this string concatenation will add up.